review/npm.md

```diff
@@ -215,10 +240,28 @@ run.
 1. To add a dependency to a manifest file, ***locally*** run [`npm
    install --save <dep-name>`](https://docs.npmjs.com/cli/v8/commands/npm-install).
 
-1. In automated environments (e.g., in CI or production) or when building
-   artifacts for end-users (e.g., container images, etc):
+1. If a project is a library, an `npm-shrinkwrap.json` should ***not*** be published. 
+   The reasoning is that version resolution should be left to the package consumer. 
+   Allow all versions from the minimum to the latest you support, e.g.,
+   `^m.n.o` to pin to a major range; `~m.n.o` to pin to a minor range. Avoid versions
+   with critical vulnerabilities as much as possible. Visit the [semver
+   calculator](https://semver.npmjs.com/) to help you define the ranges.
 
-    1. Always generate a package-lock.json ***locally*** and commit it to the
+1. If a project is a standalone CLI, developers may publish an `npm-shrinkwrap.json`. 
+   Remember that, by declaring an `npm-shrinkwrap.json`, you take responsibility 
+   for updating all the dependencies in time. Your users will not be able 
+   to update them. If you expect your CLI to be used by other projects and defined
+   in their package.json or lockfile, do **not** use `npm-shrinkwrap.json` because it will
+   hinder dependency resolution for your consumers.
+
+1. Projects that do no use a `npm-shrinkwrap.json` (libraries, standalone CLIs
+   or application projects) should declare and commit a `package-lock.json` (or other dev-only lockfile) to their repository. 
```

Comment:
This is not the best practice, similar to the shrinkwrap discussion, it depends on the type of project. So *only* applications should define a `package-lock.json` and commit it to their repos.

If libraries are using lockfiles they can accidentally publish versions which were only tested with outdated dependencies, so end users might break when they do a fresh install. The goal for library authors should be to have a CI system which does a "fresh" install so that a new user that just runs `npm i your-lib` will atleast at the time of publishing get the same experience you did in your tests.

For comparison, an application should use a lock file because for your users to "get the same experience as when authored" you need even rebuilds of the same source to produce consistent results. So the lockfile means a build from the same commit will get the same experience you intended at the time of publishing.

Reply:
you're right I botched the part on libraries. I was trying to say the following:

1.  A package-lock.json should be used and committed to the repo for all projects (reasoning: protect your developers' machines). 
2. Libraries should always ignore the lock file when running tests, and pay attention to running these tests in a non-privileged environment, e.g. via `npm install --no-package-lock`. **The problem is that many developers will run test locally anyway**
3. Applications should honor the lock file during tests, via `npm ci`

I'm trying to reconcile the comments in https://github.com/ossf/package-manager-best-practices/issues/24 (lockfiles to mitigate malicious dependencies on dev machines) and the fact that it should not be used on libraries (or in projects that `end up in other projects' manifest / lockfiles`).

Looking forward to your feedback / suggestion

Reply:
friendly ping for feedback

Reply:
Not sure where this captures the actual legitimate use of shrinkwrap files, which is when building and shipping standalone CLIs.

Reply:
the text above was just a diff from the PR's content. shrinkwrap was already discussed. Can you take a look at the section `If a project is a standalone CLI:`?

Reply:
Aaa yes I see it there now. Sorry, jumped too fast on this :-)

Reply:
@wesleytodd do you have further feedback? Are you OK with the changes?

Reply:
Really sorry for dropping off here, I totally missed the notification.

I think what landed in the "recommendations" section is technically correct, so glad to see it got sorted out. One thing I do feel here is that it is a quite complicated set of recommendations. I sent the link to a team member at work and they basically just said "I will just remove the package-lock". I think this will be a common response because the guide is technically correct at the cost of being very complicated.

Anyway, like you said below:

> I think developers will do in practice regardless of what we say (e.g. when troubleshooting/repdoducing a problem). I think the intention is to raise awareness and encourage users to run these non-locked tests in CI/less-privileged env when possible.

👍

Reply:
I don't disagree that we could make the recommendation simpler. Do you think we should include more examples, and possibly links to an example repository which contains lock files, a workflows example, etc?